server/src/main/java/org/apache/druid/metadata/SQLMetadataSegmentManager.java

```diff
@@ -83,53 +87,140 @@
 
   /**
    * Used to ensure that {@link #poll()} is never run concurrently. It should already be so (at least in production
    * code), where {@link #poll()} is called only from the task created in {@link #createPollTaskForStartOrder} and is
    * scheduled in a single-threaded {@link #exec}, so this lock is an additional safety net in case there are bugs in
    * the code, and for tests, where {@link #poll()} is called from the outside code.
    *
-   * Not using {@link #startStopLock}.writeLock() in order to still be able to run {@link #poll()} concurrently with
-   * {@link #isStarted()}.
+   * Not using {@link #startStopPollLock}.writeLock() in order to still be able to run {@link #poll()} concurrently
+   * with {@link #isPollingDatabasePeriodically()}.
    */
   private final Object pollLock = new Object();
 
   private final ObjectMapper jsonMapper;
-  private final Supplier<MetadataSegmentManagerConfig> config;
+  private final Duration periodicPollDelay;
   private final Supplier<MetadataStorageTablesConfig> dbTables;
   private final SQLMetadataConnector connector;
 
-  // Volatile since this reference is reassigned in "poll" and then read from in other threads.
-  // Starts null so we can differentiate "never polled" (null) from "polled, but empty" (empty dataSources map and
-  // empty overshadowedSegments set).
-  // Note that this is not simply a lazy-initialized variable: it starts off as null, and may transition between
-  // null and nonnull multiple times as stop() and start() are called.
-  @Nullable
-  private volatile DataSourcesSnapshot dataSourcesSnapshot = null;
+  /**
+   * This field is made volatile to avoid "ghost secondary reads" that may result in NPE, see
+   * https://github.com/code-review-checklists/java-concurrency#safe-local-dcl (note that dataSourcesSnapshot resembles
+   * a lazily initialized field). Alternative is to always read the field in a snapshot local variable, but it's too
+   * easy to forget to do.
+   *
+   * This field may be updated from {@link #exec}, or from whatever thread calling {@link #doOnDemandPoll} via {@link
+   * #awaitOrPerformDatabasePoll()} via one of the public methods of SqlSegmentsMetadata.
+   */
+  private volatile @MonotonicNonNull DataSourcesSnapshot dataSourcesSnapshot = null;
 
   /**
-   * The number of times this SQLMetadataSegmentManager was started.
+   * The latest {@link DatabasePoll} represent {@link #poll()} calls which update {@link #dataSourcesSnapshot}, either
+   * periodically (see {@link PeriodicDatabasePoll}, {@link #startPollingDatabasePeriodically}, {@link
+   * #stopPollingDatabasePeriodically}) or "on demand" (see {@link OnDemandDatabasePoll}), when one of the methods that
+   * accesses {@link #dataSourcesSnapshot}'s state (such as {@link #getImmutableDataSourceWithUsedSegments}) is
+   * called when the Coordinator is not the leader and therefore SqlSegmentsMetadata isn't polling the database
```

Comment:
I don't understand this explanation. All API calls to non-leaders should be redirected to the leader. How could database polling happen for non-leaders?

Reply:
"All API calls to non-leaders should be redirected to the leader" - you mean internally, within a Druid cluster, by a non-leader Coordinator, or externally, API callers should subscribe to the leader updates in ZK and always call to the leading Coordinator?

Reply:
All methods in `SQLMetadataSegmentManager` are called by `DruidCoordinator` or `MetadataResource`. `DruidCoordinator` has its lifecycle and doesn't do any coordinator work if it's leading. 

All HTTP calls to `MetadataResource` of non-leader from outside Druid should be [automatically redirected to the leader](https://github.com/apache/incubator-druid/blob/master/services/src/main/java/org/apache/druid/cli/CoordinatorJettyServerInitializer.java#L131).

Reply:
Interesting, thanks. I didn't know about this redirect.

It still seems to me that `SQLMetadataSegmentManager` can be called from `MetadataResource` outside of leadership between the moment when this Coordinator becomes a leader (`DruidLeaderSelector.getCurrentLeader()` returns this Coordinator's id) and `SQLMetadataSegmentManager` gets notified about that in `DruidCoordinator.becomeLeader()`. (See also the Javadoc for `LeaderLatchListener` which describes another intersting race condition.)

If this is true, I think it's worth reflecting in some comments in code. I will think about where to put this.

Reply:
Good point and the javadoc of `LeaderLatchListener` does sound interesting.

> It still seems to me that `SQLMetadataSegmentManager` can be called from `MetadataResource` outside of leadership between the moment when this Coordinator becomes the leader (`DruidLeaderSelector.getCurrentLeader()` returns this Coordinator's id) and `SQLMetadataSegmentManager` gets notified about that in `DruidCoordinator.becomeLeader()`.

My feeling about this is `DruidCoordinator.isLeader()` and `DruidCoordinator.getCurrentLeader()` could be wrong. Even when `DruidLeaderSelector` says this coordinator is leading now, `DruidCoordinator.isLeader()` shouldn't return true until `DruidCoordinator.becomeLeader()` is finished. A similar issue was in the overlord and now it fixed as [here](https://github.com/apache/incubator-druid/blob/master/indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskMaster.java#L222) (`TaskMaster.getCurrentLeader()` looks still wrong though).